iina/Logger.swift

```diff
@@ -0,0 +1,148 @@
+    let sessionDirName = "\(y)-\(m)-\(d)-\(h)-\(mm)-\(s)_\(token)"
+    let sessionDir = Utility.logDirURL.appendingPathComponent(sessionDirName, isDirectory: true)
+    Utility.createDirIfNotExist(url: sessionDir)
+    return sessionDir
+  }()
+
+  private static var logFileHandle: FileHandle = {
+    let logFileURL = logDirectory.appendingPathComponent("iina.log")
+    FileManager.default.createFile(atPath: logFileURL.path, contents: nil, attributes: nil)
+    return try! FileHandle(forWritingTo: logFileURL)
+  }()
+
+  private static var dateFormatter: DateFormatter = {
+    let formatter = DateFormatter()
+    formatter.dateFormat = "dd HH:mm:ss.SSS"
+    return formatter
+  }()
+
+  static func getLogger(_ label: String) -> Logger? {
+    return enabled ? Logger(label: label) : nil
+  }
+
+  static func closeLogFile() {
+    guard Logger.enabled else { return }
+    logFileHandle.closeFile()
+  }
+
+  private static func log(_ message: String, label: String, level: LogLevel, useNSLog: Bool, appendNewlineAtTheEnd: Bool) {
+    let time = dateFormatter.string(from: Date())
+    let string = "\(time) [\(label)][\(level.stringValue)] \(message)\(appendNewlineAtTheEnd ? "\n" : "")"
+    if useNSLog {
+      NSLog("%@", string)
+      return
+    }
+    if let data = string.data(using: .utf8) {
+      logFileHandle.write(data)
+    } else {
+      NSLog("%@", "Cannot encode log string!")
+    }
+  }
+
+  static let general: Logger? = {
+    #if DEBUG
+      let logger = Logger(label: "iina")
+      if !Logger.enabled {
+        logger.useNSLog = true
+      }
+      return logger
+    #else
+      return Logger.getLogger("iina")
+    #endif
+  }()
+
+  private init(label: String, logLevel: LogLevel? = nil) {
+    self.label = label
+    if let level = logLevel {
+      self.level = level
+    } else {
+      self.level = LogLevel(rawValue: Preference.integer(for: .logLevel).clamped(to: 0...3))!
+    }
+  }
+
+  func verbose(_ message: String, appendNewline: Bool = true) {
+    guard level <= .verbose else { return }
+    Logger.log(message, label: label, level: .verbose, useNSLog: useNSLog, appendNewlineAtTheEnd: appendNewline)
+  }
+
+  func debug(_ message: String, appendNewline: Bool = true) {
+    guard level <= .debug else { return }
+    Logger.log(message, label: label, level: .debug, useNSLog: useNSLog, appendNewlineAtTheEnd: appendNewline)
+  }
+
+  func warning(_ message: String, appendNewline: Bool = true) {
+    guard level <= .warning else { return }
+    Logger.log(message, label: label, level: .warning, useNSLog: useNSLog, appendNewlineAtTheEnd: appendNewline)
+  }
+
+  func error(_ message: String, appendNewline: Bool = true) {
+    guard level <= .error else { return }
+    Logger.log(message, label: label, level: .error, useNSLog: useNSLog, appendNewlineAtTheEnd: appendNewline)
+  }
+
+  static func assert(_ expr: Bool, _ errorMessage: String, _ block: () -> Void = {}) {
```

Comment:
`func ensure(_ condition: @autoclosure () -> Bool, _ errorMessage: String = "Assertion failed in \(#line):\(#file)", _ cleanup: () -> Void = {})`